File: electro/inline_fonts.py
#!/usr/bin/env python

from pathlib import Path
import re


def make_html_fonts_inline(path_file_in, path_file_out):
    """
    Takes an HTML file and writes a new version with inline Base64 encoded
    fonts.
    """
    path_base = path_file_in.parent
    out_lines = []
    with open(path_file_in, "r") as file:
        in_lines = [line.strip('\n') for line in file.readlines()]
    for line in in_lines:
        if 'format("woff")' in line:
            out_lines.append(convert_font(path_base, line, "woff"))
        elif 'format("woff2")' in line:
            out_lines.append(convert_font(path_base, line, "woff2"))
        else:
            out_lines.append(line)

    with open(path_file_out, 'w') as file:
        file.write('\n'.join(out_lines))


def convert_font(path_base, line, format):
    url = re.findall(r'url\("(.*?)"\)', line)
    url_expression = re.findall(r'url\(".*?"\)', line)
    if len(url) != 1 or len(url_expression) != 1:
        print(
            f'🔴 Expected to find 1 and only 1 font url entry on line: "{line}". '
            'Skipping font inline conversion.'
        )
        return line
    # Extract the (single) regex search results
    url = url[0]
    url_expression = url_expression[0]

    if url.startswith('/'):
        # Strip the leading url, so we can use it as a local path
        url = url[1:]

    path_font = path_base / Path(url)
    if not path_font.exists() and path_font.is_file():
        print(f'🔴 No font file found at url {url}. Skipping font inline conversion.')
        return line
    font_base64 = file_to_base64(path_font)
    url_inline = f'url(data:application/x-font-{format};charset=utf-8;base64,{font_base64})'

    return line.replace(url_expression, url_inline)


def file_to_base64(path_file):
    """
    Returns the content of a file as a Base64 encoded string.
    :param path_file: Path to the file.
    :type path_file: str
    :return: The file content, Base64 encoded.
    :rtype: str
    """
    import base64

    with open(path_file, 'rb') as f:
        encoded_str = base64.b64encode(f.read())
    return encoded_str.decode('utf-8')

File: pkg/electro/inlineFonts.go
package electro

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// makeHTMLFontsInline takes an HTML file and writes a new version with
// inline Base64 encoded fonts
func makeHTMLFontsInline(inFilepath, outFilepath string) error {
	var err error
	// Get the base directory of the input file
	basepath := filepath.Dir(strings.TrimRight(inFilepath, string(os.PathSeparator)))

	woffRe := regexp.MustCompile(`.*format\(["\']woff["\']\)`)
	woff2Re := regexp.MustCompile(`.*format\(["\']woff2["\']\)`)

	file, err := os.Open(inFilepath)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return err
	}

	linesOut := []string{}

	for _, line := range lines {
		if woffRe.MatchString(line) {
			// fmt.Printf("🟣 WOFF:%s\n", line)
			line, err = convertFont(basepath, line, "woff")
			if err != nil {
				return fmt.Errorf("failed to inline woff font: %w", err)
			}
			// fmt.Printf("    Converted: %s\n", line)
		}
		if woff2Re.MatchString(line) {
			// fmt.Printf("🟣 WOFF2:%s\n", line)
			line, err = convertFont(basepath, line, "woff2")
			if err != nil {
				return fmt.Errorf("failed to inline woff2 font: %w", err)
			}
			// fmt.Printf("    Converted: %s\n", line)
		}
		linesOut = append(linesOut, line)
	}

	out := strings.Join(linesOut, "\n")

	err = os.WriteFile(outFilepath, []byte(out), 0644)
	if err != nil {
		return fmt.Errorf("failed to write output file: %w", err)
	}

	return nil
}

func convertFont(basepath, line, format string) (string, error) {
	urlRe := regexp.MustCompile(`url\(["\'](.*?)["\']\)`)
	urlExpressionRe := regexp.MustCompile(`url\(["\'].*?["\']\)`)
	urls := urlRe.FindAllStringSubmatch(line, -1)
	urlExpressions := urlExpressionRe.FindAllString(line, -1)
	if len(urls) != 1 || len(urlExpressions) != 1 {
		return "", fmt.Errorf("expected to find 1 and only 1 font URL entry on line: %q", line)
	}
	url := urls[0][1]
	urlExpression := urlExpressions[0]
	// fmt.Printf("🟣 %q :: %q\n", url, urlExpression)

	url = strings.TrimPrefix(url, "/")
	pathFont := basepath + "/" + url
	if !pathExists(pathFont) || !pathIsFile(pathFont) {
		return "", fmt.Errorf("no font file found at url: %q", pathFont)
	}
	fontBase64, err := fileToBase64(pathFont)
	if err != nil {
		return "", err
	}
	urlInline := fmt.Sprintf("url(data:application/x-font-%s;charset=utf-8;base64,%s)",
		format, fontBase64)
	line = strings.Replace(line, urlExpression, urlInline, 1)

	return line, nil
}
